### ISSUE TITLE
Accessing layers not returned by the forward function of a module --or-- returning multiple layer-refs

### ISSUE THREAD

**Issue by Atticus1806** [rwth-i6/returnn_common#7]
Is there currently a way to access layers (or rather the corresponding `LayerRef`) of a Layer used in a Module but not returned by the forward function. I want to keep my Modules as generic as possible, always returning more than one layer there seems too much for me, if its just used in one or two settings.

**Comment by albertz** [comment]
No.

Just like we want to add `LayerRef.__add__`, `__sub__` etc (so that you can write `x + y` directly), we could also add some function `LayerRef.get_sub_layer(name: str) -> LayerRef` or sth like that, to access sub layers (assuming it's inside a subnet).

However, in general, I wonder if this is a good idea to use too often. Because this would then use assumptions on the internal structure of `forward` and on the returned layer, and also the naming logic. This could easily lead to bugs, or break at some point.

Also look at PyTorch (the whole API is heavily inspired from PyTorch and its `Module`). You also cannot access any intermediate outputs in its `forward` function. If there is some sensible output, even if only rarely needed, it would still be part of the output. E.g. look at [LSTM](https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html), it returns `(output, (h_n, c_n))`, and usually you only use `output`, and often write `output, _ = LSTM(...)(x)` or so.

Instead of returning a tuple, you could also return a dict, or namedtuple, which maybe would make it more structured.

Another suggestion: Maybe you could also have an argument `return_extra_x=True` on your `forward` function, and if enabled, it additionally returns whatever you need.

**Comment by Atticus1806** [comment]
I think it makes sense what you say. Changed my code to return multiple layers according to needs.

**Comment by Atticus1806** [comment]
I have to reopen this, because now when starting to debug my first "finished" model I run into problems while returning multiple layers. Maybe I am doing something wrong, so I am trying to show all the steps I am doing:
Lets assume I have a very simple Module: 
```
class EncoderConv(Module):
  def __init__(self, l2=1e-07, n_out=256, dropout=0.5):
    super().__init__()
    self.conv_0 = ConvBlock(l2=l2, n_out=n_out, dropout=dropout)
    self.conv_1 = ConvBlock(l2=l2, n_out=n_out, dropout=dropout)
    self.conv_2 = ConvBlock(l2=l2, n_out=n_out, dropout=dropout)

  def forward(self, embedding: LayerRef) -> LayerRef:
    x_ = self.conv_0(embedding)
    x = self.conv_1(x_)
    x = self.conv_2(x)
    return x
```
Now I want to return `x_` aswell (for some reason), so I change the last line to: `return x, x_`. What currently then happens is, that it creates an output layer of class copy which gets both as from parameters (even with a tuple instead of list). Then I changed the `make_layer_dict` function, to check wether res is a Layer or a Tuple. If Layer act as before, if Tuple use the first entry as input to the copy layer (which could be a list if we want to concat multiple layer entries). I can make a PR of this once we found a solution ofc. The problem is now, that this does not work as expected:
```
# in init
self.conv = EncoderConv()

# in forward
x, x_ = self.conv(some input)
``` 
I still get just the network dict as a return (with fixed copy output layer at the end) but ofc. this reasults in an unpacking error and also is not what I want to recieve. I hope the problem is somewhat clear.  I am quite sure that I am doing something wrong, I am just not sure what that is.

**Comment by albertz** [comment]
You are right, this is not implemented yet. Maybe I should have started to implement this right from the beginning. But it should be hard to add now.

We have to make consistent usage of `nest`. You already find some usage of it. You find very similar code and handling also in our pytorch-to-returnn repo.

I don't exactly understand what the problem is now (except that this is not implemented yet, and needs to be implemented).

**Comment by albertz** [comment]
Ok, while looking a bit through the code, I think I see why it is not so straightforward.

We either need to extend `ILayerMaker.make_layer_dict` or extend `ILayerMaker` in some other way (or maybe just `ISubnet`).

I'm currently thinking about extending `ILayerMaker.make_layer_dict` to return type:
`Union[LayerDictRaw, Tuple[LayerDictRaw, Any]`.

So either just as before, and in the second case, this would be used by `ISubnet.make_layer_dict`. When it calls `res = self._subnet_func(*args, **kwargs)`, it has to check whether `res` is sth more complex, and then return layer refs for the other args, in the same structure as `res` (you can use `nest` to replicate the same structure).

**Comment by Atticus1806** [comment]
Yes this is what I was thinking of too, that is kind of the distinction I mentioned I already did even though its not really finished yet. 
We could check wether the return of the forward is of instance either `Layer` or `List[Layer]` and then handle it as input for the `Copy` layer and otherwise do something like `assert(res, Tuple)` just to check reasonability and then add `Copy()(res[0], name="output)` and fetch the `LayerRef` of the other layers in that are in `res` and return them all together. I am not sure what you have done yet, those are just my thoughts, I can try to put them into a PR later when I get time to, but if you are already doing something just keep on I guess.

**Comment by Atticus1806** [comment]
Hmm I wrote a hacky first version, but now I realize that this would also need changes to the `call` function of the `ILayerMaker` which might turn out a bit more complicated.

**Comment by albertz** [comment]
Most of this is pretty straightforward and all automatic with `nest`. See maybe also related code in pytorch-to-returnn.

**Comment by Atticus1806** [comment]
After reading into the building code and trying to understand what is happening there, I made some changes which I pushed to [benedikt_mul_returns](https://github.com/rwth-i6/returnn_common/tree/benedikt_mul_returns)
I have other stuff to do for now and in my Modules I do recieve an error in a line for which I cannot say right now wether its related to what I changed or if its something unrelated. (Maybe you can quickly see wether it is caused by my changes so this is the line where it crashes):
`x = self.generic(None, base=encoder_concat, weights=weights, name="generic` , the issue seems to be related to the None as inputs

I need to stop for now but also think I kind of hit a wall, so maybe you can have a look at the PR and comment wether it might work (similar) to this.

**Comment by albertz** [comment]
I cannot really tell without context about the problem with your `self.generic` call. It looks a bit strange to me that you pass `None` there. Also, you never ever should need to explicitly set `name`.

But maybe it's also related to the problems I commented in the PR. For the PR, let's also create a very simple test case (maybe just `def forward(self, x): return x, x` or so).

**Comment by Atticus1806** [comment]
>Also, you never ever should need to explicitly set name.

You are right, this was more of a habit I kept doing when picking up using `returnn_common` and was still in the example since I just copy pasted a really "early" example.

>It looks a bit strange to me that you pass None there.

For now I am "only" trying to copy a returnn config into the new style. In the config the layer (after some subnetwork transformations) looks like this:
`'0': {'base': 'base:base:encoder/concat', 'class': 'generic_attention', 'weights': 'weights'},` which does not have an explicit "from",  but since in the current version you need to set `from:` I decided on `None` but maybe again this is a change we would make to the layer generation instead. If this error or weird behavior persists even after this Issue/PR I will make a new Issue.

**Comment by albertz** [comment]
> `GenericAttention` does not have an explicit "from", but since in the current version you need to set `from:` I decided on `None` but maybe again this is a change we would make to the layer generation instead.

Ah yes, sure. Yes, this should be fixed in the layer generation. I just fixed this via 0e8a118f5f60e5b163e7d91.

**Comment by albertz** [comment]
I think this got fixed via #13 (831b106), right? So closing now.
